<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analysis Tools for TIRTLseq Data • TIRTLtools</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analysis Tools for TIRTLseq Data">
<meta name="description" content="A suite of tools for analyzing paired T-cell receptor (TCR) datasets created using TIRTL-seq. Contains an efficient GPU implementation of TCRdist (Dash et al., Nature 2017).">
<meta property="og:description" content="A suite of tools for analyzing paired T-cell receptor (TCR) datasets created using TIRTL-seq. Contains an efficient GPU implementation of TCRdist (Dash et al., Nature 2017).">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="tirtltools-r-package">TIRTLtools R package<a class="anchor" aria-label="anchor" href="#tirtltools-r-package"></a>
</h1></div>
<p><strong><em>Package is in active development and may change frequently.</em></strong></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>TIRTLtools</code> is a suite of tools for analyzing T-cell receptor (TCR) repertoires created using <strong>TIRTL-seq</strong> (<strong>T</strong>hroughput-<strong>I</strong>ntensive <strong>R</strong>apid <strong>T</strong>CR <strong>L</strong>ibrary <strong>seq</strong>uencing) <a href="https://www.biorxiv.org/content/10.1101/2024.09.16.613345v2" class="external-link">(Pogorelyy and Kirk et al., bioRxiv 2024)</a>. We provide functions for analysis of <strong>paired TCR repertoires</strong> as well as <strong>single-chain bulk data</strong>.</p>
<p>In addition to various <strong>analysis and plotting functions</strong>, we provide an <strong>efficient batched GPU implementation of TCRdist</strong> (<a href="https://doi.org/10.1038/nature22383" class="external-link">Dash et al., Nature 2017</a>) that works with both NVIDIA and Apple Silicon GPUs. In testing, we were able to calculate pairwise TCRdist for a repertoire of ~1 million TCRs in a few hours using a MacBook Pro (16-core GPU, M4 Pro).</p>
<div class="section level3">
<h3 id="input-data-for-package">Input data for package<a class="anchor" aria-label="anchor" href="#input-data-for-package"></a>
</h3>
<p>The package requires paired-chain and single-chain “pseudo-bulk” data files created by the <a href="https://github.com/pogorely/TIRTL" class="external-link">TIRTLseq pipeline</a>.</p>
<p>The pipeline reads <a href="https://mixcr.com/" class="external-link">MiXCR</a> output files from TCR repertoire sequencing for individual wells in a multiwell plate (e.g. 96-well or 384-well plate). It uses two complementary algorithms, <a href="https://doi.org/10.1093/bioinformatics/bty801" class="external-link">MAD-HYPE</a> and <a href="https://www.biorxiv.org/content/10.1101/2024.09.16.613345v2" class="external-link">T-SHELL</a>, to pair alpha and beta chains. It outputs three files: a .tsv file containing alpha and beta chain pairs predicted by each algorithm (file ending in “TIRTLoutput.tsv”), and two .tsv files containing aggregated single-chain “pseudo-bulk” data for alpha or beta chains from all wells on a plate (files ending in “pseudobulk_TRA.tsv” and “pseudobulk_TRB.tsv”).</p>
<p><strong>MAD-HYPE</strong> (<strong>M</strong>ulticell <strong>A</strong>nalytical <strong>D</strong>econvolution for <strong>H</strong>igh <strong>Y</strong>ield <strong>P</strong>aired-chain <strong>E</strong>valuation, Holec and Berleant et al., 2019) is a Bayesian method that computes the likelihood of alpha and beta chain matches based on their co-occurence patterns in wells on a plate. We re-implemented the MAD-HYPE algorithm for GPU computation, increasing efficiency and speed by two orders of magnitude.</p>
<p><strong>T-SHELL</strong> (<strong>T</strong>CR ɑβ <strong>S</strong>equence <strong>H</strong>ighly <strong>E</strong>fficient <strong>L</strong>inkage <strong>L</strong>earning, Pogorelyy and Kirk et al., bioRxiv 2024) is an algorithm we developed that uses the correlation between TCRalpha and TCRbeta clonotype relative frequencies across wells in a plate (rather than their presence or absence) to pair them. It works well in the case of the most abundant clones, which are found in all or almost all wells, making it impossible for the MAD-HYPE algorithm to pair them using their co-occurrence patterns.</p>
<p>Example input data for the TIRTLseq pipeline can be found on <a href="https://zenodo.org/records/14010377" class="external-link">Zenodo</a>. For example, the aph17pre_exp_clones.tar.gz (172 MB) and aph17post_exp_clones.tar.gz (1.1 GB) files contain well-level TCR clonotype data (MiXCR output) for pre- and post-antigen-independent T cell expansion experiments.</p>
</div>
<div class="section level3">
<h3 id="package-functions">Package functions<a class="anchor" aria-label="anchor" href="#package-functions"></a>
</h3>
<div class="section level4">
<h4 id="data-loading-and-filtering">Data loading and filtering:<a class="anchor" aria-label="anchor" href="#data-loading-and-filtering"></a>
</h4>
<ul>
<li>
<code><a href="reference/load_tirtlseq.html">load_tirtlseq()</a></code> – load paired TCR and single-chain data from TIRTL-seq experiments</li>
<li>
<code><a href="reference/filter_dataset.html">filter_dataset()</a></code> – subset a TIRTLseq dataset to only the requested samples</li>
</ul>
</div>
<div class="section level4">
<h4 id="repertoire-analysis">Repertoire analysis:<a class="anchor" aria-label="anchor" href="#repertoire-analysis"></a>
</h4>
<ul>
<li>
<code><a href="reference/TCRdist.html">TCRdist()</a></code> – calculate TCRdist via batched GPU calculation</li>
<li>
<code><a href="reference/cluster_tcrs.html">cluster_tcrs()</a></code> – cluster TCRs (with themselves and with TCRs from VDJ-db) based on pairwise TCRdist</li>
<li>
<code>diversity()</code> – calculate repertoire diversity metrics</li>
</ul>
</div>
<div class="section level4">
<h4 id="plotting">Plotting:<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h4>
<ul>
<li>
<code>sample_overlap()</code> – plot overlap/agreement between samples (in terms of most frequent clones)</li>
<li>
<code><a href="reference/plot_sample_vs_sample.html">plot_sample_vs_sample()</a></code> – scatterplot of readFraction between two samples</li>
<li>
<code>rank_plot()</code> – a line plot of clonotype rank vs. readFraction for different samples</li>
<li>
<code><a href="reference/plot_clusters.html">plot_clusters()</a></code> – plot top “n” clusters with the most observed TCRs – currently outputs a network graph, a umap, and a heatmap</li>
<li>
<code>plot_gene_usage()</code> – bar plots of V-gene segment usage (or J-gene segment usage)</li>
<li>
<code><a href="reference/plot_num_partners.html">plot_num_partners()</a></code> – plot number of partners for alpha and beta chains in paired data</li>
<li>
<code><a href="reference/plot_clonotype_indices.html">plot_clonotype_indices()</a></code> – stacked bar chart with fractions of reads for most frequent clonotypes.</li>
<li>
<code><a href="reference/plot_diversity.html">plot_diversity()</a></code> – bar plots of different diversity metrics</li>
</ul>
</div>
<div class="section level4">
<h4 id="helper-functions">Helper functions:<a class="anchor" aria-label="anchor" href="#helper-functions"></a>
</h4>
<ul>
<li>
<code><a href="reference/add_single_chain_data.html">add_single_chain_data()</a></code> – add alpha- and beta-chain read counts/fractions to the paired TCR data</li>
<li>
<code><a href="reference/identify_non_functional_seqs.html">identify_non_functional_seqs()</a></code> – mark cdr3 sequences that have stop codons (*) or frame shifts (_)</li>
<li>
<code><a href="reference/identify_paired.html">identify_paired()</a></code> – identify which alpha and beta chains in the pseudo-bulk data are found in pairs</li>
<li>
<code><a href="reference/get_all_div_metrics.html">get_all_div_metrics()</a></code> – returns all options for diversity metrics (currently: simpson, gini, gini.simpson, inv.simpson, shannon, berger.parker, richness, d50, dXX, renyi, hill)</li>
<li>
<code><a href="reference/get_all_tcrs.html">get_all_tcrs()</a></code> – aggregate all of the TCRs from all samples in a dataset.</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the package directly from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"NicholasClark/TIRTLtools"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you may clone the repository to a local folder</p>
<pre><code>git clone https://github.com/NicholasClark/TIRTLtools.git</code></pre>
<p>and load it using devtools:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">load_all</span><span class="op">(</span><span class="st">"./TIRTLtools"</span><span class="op">)</span> <span class="co">### location of cloned repo on your machine</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tcrdist-gpu-implementation">TCRdist GPU implementation<a class="anchor" aria-label="anchor" href="#tcrdist-gpu-implementation"></a>
</h2>
<p>This function computes the similarity of two T-cell receptors (TCRs) using TCRdist (<a href="https://doi.org/10.1038/nature22383" class="external-link">Dash et al., Nature 2017</a>). The function can efficiently calculate pairwise similarities of thousands or millions of TCRs using a GPU if it is available. It relies on python code using <a href="https://cupy.dev/" class="external-link">cupy</a> (Nvidia GPUs) or <a href="https://opensource.apple.com/projects/mlx/" class="external-link">mlx</a> (Apple Silicon GPUs), which is called through the <code>reticulate</code> R package.</p>
<p>The following example computes TCRdist values for all pairs of annotated TCRs in the <a href="https://vdjdb.cdr3.net/" class="external-link">VDJdb</a> database. It outputs all values with TCRdist &lt;= 90.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#library(TIRTLtools)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">load_all</span><span class="op">(</span><span class="st">"./TIRTLtools"</span><span class="op">)</span> <span class="co">### location of cloned repo on your machine</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Load an example dataset (annotated TCRs from VDJ-db)</span></span>
<span><span class="va">tcr1</span> <span class="op">=</span> <span class="fu">TIRTLtools</span><span class="fu">::</span><span class="va">vdj_db</span></span>
<span><span class="co">### note: You may replace this with a file of your choice - needs to have columns "va", "vb", "cdr3a", and "cdr3b"</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">tcr1</span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 26,213 × 15</span></span>
<span><span class="co">#       complex.id cdr3a         va       ja    cdr3b vb    jb    Species MHC.A MHC.B MHC.class Epitope Epitope.gene</span></span>
<span><span class="co">#         &lt;int&gt;   &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;</span></span>
<span><span class="co"># 1          5 CAVRDGGTGFQKLVF   TRAV3*01 TRAJ… CASR… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      HPVGEA… EBNA1</span></span>
<span><span class="co"># 2          6 CAARGIGSGTYKYIF   TRAV13-… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      HPVGEA… EBNA1</span></span>
<span><span class="co"># 3          8 CAVRDGTANNLFF     TRAV3*01 TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 4          9 CAFMKDAGGTSYGKLTF TRAV38-… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 5         10 CAGGGSQGNLIF      TRAV27*… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 6         11 CALAGSQGNLIF      TRAV6*01 TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 7         12 CAGADGGSQGNLIF    TRAV27*… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 8         13 CAAGGSQGNLIF      TRAV27*… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 9         14 CAAPPSNTGKLIF     TRAV3*01 TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 10         15 CAVRDLLTNAGKSTF   TRAV3*01 TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># # ℹ 26,203 more rows</span></span>
<span><span class="co"># # ℹ 2 more variables: Epitope.species &lt;chr&gt;, Reference &lt;chr&gt;</span></span>
<span><span class="co"># # ℹ Use `print(n = ...)` to see more rows</span></span>
<span><span class="va">result</span> <span class="op">=</span> <span class="fu"><a href="reference/TCRdist.html">TCRdist</a></span><span class="op">(</span>tcr1 <span class="op">=</span> <span class="va">tcr1</span>, tcrdist_cutoff <span class="op">=</span> <span class="fl">90</span>, chunk_size <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="co"># Checking for available GPU...</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Apple Silicon GPU detected:</span></span>
<span><span class="co">#   Apple Silicon GPU (M1/M2/M3)</span></span>
<span><span class="co"># Checking for GPU-related Python modules...</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 'mlx' is installed (for Apple Silicon GPUs).</span></span>
<span><span class="co"># Loading mlx to perform TCRdist</span></span>
<span><span class="co"># total number of chunks (rows): 5</span></span>
<span><span class="co"># total number of chunks (cols): 5</span></span>
<span><span class="co"># Processing chunk (rows) 0</span></span>
<span><span class="co"># Processing chunk (rows) 5000</span></span>
<span><span class="co"># Processing chunk (rows) 10000</span></span>
<span><span class="co"># Processing chunk (rows) 15000</span></span>
<span><span class="co"># Processing chunk (rows) 20000</span></span>
<span><span class="co"># Processing chunk (rows) 25000</span></span>
<span><span class="co"># Time taken: 2.970693 seconds</span></span>
<span></span>
<span><span class="va">edge_df</span> <span class="op">=</span> <span class="va">result</span><span class="op">[[</span><span class="st">'TCRdist_df'</span><span class="op">]</span><span class="op">]</span> <span class="co">### table of TCRdist values &lt;= cutoff</span></span>
<span><span class="va">meta_df</span> <span class="op">=</span> <span class="va">result</span><span class="op">[[</span><span class="st">'tcr1'</span><span class="op">]</span><span class="op">]</span> <span class="co">### table of input data with indices</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">edge_df</span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 101,797 × 3</span></span>
<span><span class="co">#       edge1_0index edge2_0index TCRdist</span></span>
<span><span class="co">#           &lt;int&gt;        &lt;int&gt;   &lt;int&gt;</span></span>
<span><span class="co"># 1            5            4      87</span></span>
<span><span class="co"># 2            7            4       3</span></span>
<span><span class="co"># 3            7            5      90</span></span>
<span><span class="co"># 4           12            4      15</span></span>
<span><span class="co"># 5           12            5      78</span></span>
<span><span class="co"># 6           12            7      12</span></span>
<span><span class="co"># 7           13            4      78</span></span>
<span><span class="co"># 8           13            6      72</span></span>
<span><span class="co"># 9           13            7      81</span></span>
<span><span class="co"># 10           14            6      45</span></span>
<span><span class="co"># # ℹ 101,787 more rows</span></span>
<span><span class="co"># # ℹ Use `print(n = ...)` to see more rows</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">meta_df</span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 26,213 × 16</span></span>
<span><span class="co">#   complex.id cdr3a      va    ja    cdr3b vb    jb    Species MHC.A MHC.B MHC.class Epitope Epitope.gene</span></span>
<span><span class="co">#       &lt;int&gt;   &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;</span></span>
<span><span class="co"># 1          5 CAVRDGGTG… TRAV… TRAJ… CASR… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      HPVGEA… EBNA1</span></span>
<span><span class="co"># 2          6 CAARGIGSG… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      HPVGEA… EBNA1</span></span>
<span><span class="co"># 3          8 CAVRDGTAN… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 4          9 CAFMKDAGG… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 5         10 CAGGGSQGN… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 6         11 CALAGSQGN… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 7         12 CAGADGGSQ… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 8         13 CAAGGSQGN… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 9         14 CAAPPSNTG… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># 10         15 CAVRDLLTN… TRAV… TRAJ… CASS… TRBV… TRBJ… HomoSa… HLA-… B2M   MHCI      GILGFV… M</span></span>
<span><span class="co"># # ℹ 26,203 more rows</span></span>
<span><span class="co"># # ℹ 3 more variables: Epitope.species &lt;chr&gt;, Reference &lt;chr&gt;, tcr_index &lt;dbl&gt;</span></span>
<span><span class="co"># # ℹ Use `print(n = ...)` to see more rows</span></span>
<span></span>
<span><span class="co">#' TCRdist function output:</span></span>
<span><span class="co">#' A list with two items:</span></span>
<span><span class="co">#'    'TCRdist_df' - A dataframe containing all TCRdist values less than the cutoff (default = 90).</span></span>
<span><span class="co">#'                  Contains the indices of the two TCRs and the TCRdist value.</span></span>
<span><span class="co">#'    'tcr1' - A dataframe with the input data, along with a column "tcr_index" with the index of each TCR</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>For details on our pairing pipeline, see the <a href="https://www.biorxiv.org/content/10.1101/2024.09.16.613345v2" class="external-link">TIRTLseq preprint (Pogorelyy and Kirk et al.)</a> and our <a href="https://github.com/pogorely/TIRTL" class="external-link">github repository</a>.</p>
<p>For details on the MAD-HYPE algorithm, see <a href="https://academic.oup.com/bioinformatics/article/35/8/1318/5095649" class="external-link">Holec and Berleant et al., 2019</a>.</p>
<p>For details on MiXCR, see their <a href="https://mixcr.com/" class="external-link">website</a> and publications in <a href="https://www.nature.com/articles/nmeth.3364" class="external-link">Nature Methods (2015)</a> and <a href="https://www.nature.com/articles/nbt.3979" class="external-link">Nature Biotechnology (2017)</a>.</p>
<p>For details on TCRdist, see <a href="https://doi.org/10.1038/nature22383" class="external-link">Dash et al., Nature 2017</a></p>
<p><br></p>
<p>Written by Mikhail Pogorelyy and Nicholas Clark, <a href="https://www.stjude.org/research/labs/thomas-lab.html" class="external-link">Thomas Lab, St. Jude Children’s Research Hospital</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TIRTLtools</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Nicholas Clark <br><small class="roles"> Author, maintainer </small>   </li>
<li>Mikhail Pogorelyy <br><small class="roles"> Author </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicholas Clark, Mikhail Pogorelyy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
