<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analysis Tools for TIRTLseq Data • TIRTLtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Analysis Tools for TIRTLseq Data">
<meta name="description" content="A suite of tools for analyzing paired T-cell receptor (TCR) datasets created using the TIRTL-seq assay. Contains an efficient GPU implementation of TCRdist (Dash et al., Nature 2017).">
<meta property="og:description" content="A suite of tools for analyzing paired T-cell receptor (TCR) datasets created using the TIRTL-seq assay. Contains an efficient GPU implementation of TCRdist (Dash et al., Nature 2017).">
<meta property="og:image" content="https://nicholasclark.github.io/TIRTLtools/logo.svg">
</head>
<body>
<p>









  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="articles/TIRTLtools.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="articles/data_loading.html">Loading TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="articles/plot_examples.html">Plot examples with SJTRC COVID-19 data</a></li>
    <li><a class="dropdown-item" href="articles/TCRdist.html">Fast TCR similarity calculation with TCRdist GPU</a></li>
    <li><a class="dropdown-item" href="articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic TIRTLseq data analysis</h6></li>
    <li><a class="dropdown-item" href="articles/summarizing_diversity.html">Estimating TCR Repertoire diversity</a></li>
    <li><a class="dropdown-item" href="articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Constructing full TCR nucleotide sequences</h6></li>
    <li><a class="dropdown-item" href="articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Well-level TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.svg" class="logo" alt=""><h1 id="tirtltools-r-package">TIRTLtools R package<a class="anchor" aria-label="anchor" href="#tirtltools-r-package"></a>
</h1>
</div>
<p><strong><em>Package is in active development and may change frequently.</em></strong></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>TIRTLtools</code> is a suite of tools for analyzing T-cell receptor (TCR) repertoires created using <strong>TIRTL-seq</strong> (<strong>T</strong>hroughput-<strong>I</strong>ntensive <strong>R</strong>apid <strong>T</strong>CR <strong>L</strong>ibrary <strong>seq</strong>uencing) <a href="https://www.biorxiv.org/content/10.1101/2024.09.16.613345v2" class="external-link">(Pogorelyy and Kirk et al., bioRxiv 2024)</a>. We provide functions for analysis of <strong>paired TCR repertoires</strong> as well as <strong>single-chain bulk data</strong>.</p>
<p>In addition to various <strong>analysis and plotting functions</strong>, we provide an <strong>efficient batched GPU implementation of TCRdist</strong> (<a href="https://doi.org/10.1038/nature22383" class="external-link">Dash et al., Nature 2017</a>) that works with both NVIDIA and Apple Silicon GPUs. In testing, we were able to calculate pairwise TCRdist for a repertoire of ~1 million TCRs in a few hours using a MacBook Pro (16-core GPU, M4 Pro).</p>
</div>
<div class="section level2">
<h2 id="the-tirtlseq-assay">The TIRTLseq assay<a class="anchor" aria-label="anchor" href="#the-tirtlseq-assay"></a>
</h2>
<p><a href="https://doi.org/10.1101/2024.09.16.613345" class="external-link">TIRTL-seq</a> is a method to obtain paired TCR-sequencing information from millions of human T cells. Cells are split across wells in <a href="https://www.protocols.io/view/tirtl-seq-384-well-eq2lyx37wgx9/" class="external-link">384-well</a> or <a href="https://www.protocols.io/view/tirtl-seq-96-well-kqdg3k6xpv25/" class="external-link">96-well</a> plates with ~2,500 to ~25,000 cells per well and alpha and beta TCR chains are sequenced simultaneously in each well.</p>
<p>The initial output from the TIRTLseq assay consists of two tables of read counts for single-chain TCRs (one for alpha and one for beta) for each well. Alpha and beta chains are paired using computational algorithms based on their occurrence patterns and read counts across wells.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Instructions to <a href="articles/TIRTLtools.html">install TIRTLtools</a>.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>For details on our pairing pipeline, see the <a href="https://www.biorxiv.org/content/10.1101/2024.09.16.613345v2" class="external-link">TIRTLseq preprint (Pogorelyy and Kirk et al.)</a> and our <a href="https://github.com/pogorely/TIRTL" class="external-link">github repository</a>.</p>
<p>For details on the MAD-HYPE algorithm, see <a href="https://academic.oup.com/bioinformatics/article/35/8/1318/5095649" class="external-link">Holec and Berleant et al., 2019</a>.</p>
<p>For details on MiXCR, see their <a href="https://mixcr.com/" class="external-link">website</a> and publications in <a href="https://www.nature.com/articles/nmeth.3364" class="external-link">Nature Methods (2015)</a> and <a href="https://www.nature.com/articles/nbt.3979" class="external-link">Nature Biotechnology (2017)</a>.</p>
<p>For details on TCRdist, see <a href="https://doi.org/10.1038/nature22383" class="external-link">Dash et al., Nature 2017</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TIRTLtools</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a> <br><small class="roles"> Author, maintainer </small>   </li>
<li>
<a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a> <br><small class="roles"> Author </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  
</body>
</html>
