<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Cluster TCRs (using the Leiden algorithm) based on their pairwise TCRdist values — cluster_tcrs • TIRTLtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cluster TCRs (using the Leiden algorithm) based on their pairwise TCRdist values — cluster_tcrs"><meta name="description" content="The cluster_tcrs() function aggregates all of the paired TCRs found in the data,
calculates pairwise similarity using the va, vb, cdr3a, and cdr3b regions (via TCRdist),
and clusters the results using the Leiden algorithm."><meta property="og:description" content="The cluster_tcrs() function aggregates all of the paired TCRs found in the data,
calculates pairwise similarity using the va, vb, cdr3a, and cdr3b regions (via TCRdist),
and clusters the results using the Leiden algorithm."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Cluster TCRs (using the Leiden algorithm) based on their pairwise TCRdist values</h1>

      <div class="d-none name"><code>cluster_tcrs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The <code>cluster_tcrs()</code> function aggregates all of the paired TCRs found in the data,
calculates pairwise similarity using the va, vb, cdr3a, and cdr3b regions (via TCRdist),
and clusters the results using the Leiden algorithm.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cluster_tcrs</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  tcrdist_cutoff <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  with_db <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  db <span class="op">=</span> <span class="fu">TIRTLtools</span><span class="fu">::</span><span class="va">vdj_db</span>,</span>
<span>  allow_self_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  remove_MAIT <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>a list of TIRTLseq TCR data for samples created with <code><a href="load_tirtlseq.html">load_tirtlseq</a>()</code></p></dd>


<dt id="arg-tcrdist-cutoff">tcrdist_cutoff<a class="anchor" aria-label="anchor" href="#arg-tcrdist-cutoff"></a></dt>
<dd><p>the <code><a href="TCRdist.html">TCRdist</a>()</code> function will only record TCRdist values
less than or equal to the cutoff. Default is 90. Note: Higher cutoffs will return more
data, at most NxN where N is the number of unique TCRs.</p></dd>


<dt id="arg-resolution">resolution<a class="anchor" aria-label="anchor" href="#arg-resolution"></a></dt>
<dd><p>the "resolution" parameter for the Leiden algorithm. A lower
value will produce larger clusters and a higher value will produce smaller clusters.
Typical values are in the 0.1 - 2.0 range. A higher value may be better for densely
connected data while a lower value may be better for moderately connected data. Default is 0.1.</p></dd>


<dt id="arg-with-db">with_db<a class="anchor" aria-label="anchor" href="#arg-with-db"></a></dt>
<dd><p>if TRUE, observed clones will be compared and clustered with a dataframe of
annotated clones. By default, a dataframe with VDJ-db annotations is used.</p></dd>


<dt id="arg-db">db<a class="anchor" aria-label="anchor" href="#arg-db"></a></dt>
<dd><p>a data frame with annotated TCRs. The default is the VDJ-db database.</p></dd>


<dt id="arg-allow-self-edges">allow_self_edges<a class="anchor" aria-label="anchor" href="#arg-allow-self-edges"></a></dt>
<dd><p>if FALSE, only calculate TCRdist between input data TCRs
and the TCR annotation database (db). If TRUE, calculate pairwise TCRdist for all
of the data including the input and the annotated TCRs.</p></dd>


<dt id="arg-remove-mait">remove_MAIT<a class="anchor" aria-label="anchor" href="#arg-remove-mait"></a></dt>
<dd><p>remove MAIT TCRs before clustering (default is TRUE)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a list with the following elements:</p>
<p><code>$df</code> - a data frame with all unique TCRs along with cluster annotations</p>
<p><code>$dist_df</code> - a data frame with distances (TCRdist) between TCR pairs in long format</p>
<p><code>$sparse_adj_mat</code> - an adjacency matrix (in sparse format) marking TCR pairs with TCRdist &lt;= tcrdist_cutoff</p>
<p><code>$graph_adj</code> - an igraph object created from the adjacency matrix</p>
<p><code>$tcrdist_cutoff</code> - the cutoff used for TCRdist</p>
<p><code>$resolution</code> - the resolution parameter used for the Leiden algorithm</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function also filters the dataset to TCRs that are valid for TCRdist.</p>
<p>The following TCRs are removed:
 - TCRs that contain stop codons (*) or frame shifts (_) in their cdr3a or cdr3b regions
 - TCRs that contain a cdr3 region with 5 or less amino acids
 - TCRs that contain a v segment allele not found in our parameter table</p>
<p>V-segments that do not specify an allele (e.g. "TRAV1-2" instead of "TRAV1-2*01")
will be assigned to the "*01" allele.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="plot_clusters.html">plot_clusters</a>()</code>, <code><a href="identify_non_functional_seqs.html">identify_non_functional_seqs</a>()</code>, <code><a href="TCRdist.html">TCRdist</a>()</code></p>
<p>Other repertoire_analysis:
<code><a href="TCRdist.html">TCRdist</a>()</code>,
<code><a href="TCRdist_cpp.html">TCRdist_cpp</a>()</code>,
<code><a href="calculate_diversity.html">calculate_diversity</a>()</code>,
<code><a href="get_all_div_metrics.html">get_all_div_metrics</a>()</code>,
<code><a href="get_all_tcrs.html">get_all_tcrs</a>()</code>,
<code><a href="get_pair_stats.html">get_pair_stats</a>()</code>,
<code><a href="get_paired_by_read_fraction_range.html">get_paired_by_read_fraction_range</a>()</code>,
<code><a href="summarize_data.html">summarize_data</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span><span class="co"># paired = load_tirtlseq("your_directory/")</span></span></span>
<span class="r-in"><span><span class="co"># obj = cluster_tcrs(paired)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

