<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ TIRTLtools</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/TIRTLtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/data_loading.html">Data loading</a></li>
    <li><a class="dropdown-item" href="../articles/plot_examples.html">Plot examples</a></li>
    <li><a class="dropdown-item" href="../articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><a class="dropdown-item" href="../articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><a class="dropdown-item" href="../articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><a class="dropdown-item" href="../articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/summarizing_diversity.html">(coming soon) Summarizing TCR diversity</a></li>
    <li><a class="dropdown-item" href="../articles/TCRdist.html">TCRdist GPU implementation</a></li>
    <li><a class="dropdown-item" href="../articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
    <li><a class="dropdown-item" href="../articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="bulk-tirtlseq-data">Bulk TIRTLseq data<a class="anchor" aria-label="anchor" href="#bulk-tirtlseq-data"></a></h2>




    </div><div class="section level2">

      <h3 id="data-loading-and-wrangling">Data loading and wrangling<a class="anchor" aria-label="anchor" href="#data-loading-and-wrangling"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_metadata.html">add_metadata()</a></code>

        </dt>
        <dd>Add metadata to a TIRTLseqData object</dd>
      </dl><dl><dt>

          <code><a href="filter_dataset.html">filter_dataset()</a></code>

        </dt>
        <dd>Subset a TIRTLseqData object</dd>
      </dl><dl><dt>

          <code><a href="load_tirtlseq.html">load_tirtlseq()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Load data from TIRTLseq experiments</dd>
      </dl><dl><dt>

          <code><a href="reorder_samples.html">reorder_samples()</a></code>

        </dt>
        <dd>Change the order of samples in a TIRTLseqData object</dd>
      </dl></div><div class="section level2">

      <h3 id="data-processing-and-cleaning">Data processing and cleaning<a class="anchor" aria-label="anchor" href="#data-processing-and-cleaning"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="TIRTL_process.html">TIRTL_process()</a></code>

        </dt>
        <dd>Run data processing functions on a TIRTLseqData object</dd>
      </dl><dl><dt>

          <code><a href="add_single_chain_data.html">add_single_chain_data()</a></code>

        </dt>
        <dd>Add single-chain read counts/fractions to the paired TCR data</dd>
      </dl><dl><dt>

          <code><a href="clean_pairs.html">clean_pairs()</a></code>

        </dt>
        <dd>Remove excess pairs for individual single chains</dd>
      </dl><dl><dt>

          <code><a href="filter_mait.html">filter_mait()</a></code>

        </dt>
        <dd>Remove MAIT (Mucosal-Associated Invariant T cells) TCRs</dd>
      </dl><dl><dt>

          <code><a href="identify_non_functional_seqs.html">identify_non_functional_seqs()</a></code>

        </dt>
        <dd>Identify TCRs that contain non-functional CDR3 sequences</dd>
      </dl><dl><dt>

          <code><a href="identify_paired.html">identify_paired()</a></code>

        </dt>
        <dd>Identify which single chains were paired</dd>
      </dl><dl><dt>

          <code><a href="prep_for_tcrdist.html">prep_for_tcrdist()</a></code>

        </dt>
        <dd>Prepare paired TCRs for TCRdist calculation</dd>
      </dl><dl><dt>

          <code><a href="remove_duplicates.html">remove_duplicates()</a></code>

        </dt>
        <dd>Removes duplicate paired TCRs</dd>
      </dl></div><div class="section level2">

      <h3 id="tcr-repertoire-analysis">TCR repertoire analysis<a class="anchor" aria-label="anchor" href="#tcr-repertoire-analysis"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="TCRdist.html">TCRdist()</a></code>

        </dt>
        <dd>GPU implementation of TCRdist, a distance/similarity metric for pairs of TCRs</dd>
      </dl><dl><dt>

          <code><a href="TCRdist_cpp.html">TCRdist_cpp()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Parallelized C++ implementation of TCRdist (no GPU required)</dd>
      </dl><dl><dt>

          <code><a href="calculate_diversity.html">calculate_diversity()</a></code> <code><a href="calculate_diversity.html">diversity()</a></code>

        </dt>
        <dd>Calculate TCR repertoire diversity metrics</dd>
      </dl><dl><dt>

          <code><a href="cluster_tcrs.html">cluster_tcrs()</a></code>

        </dt>
        <dd>Cluster TCRs (using the Leiden algorithm) based on their pairwise TCRdist values</dd>
      </dl><dl><dt>

          <code><a href="get_all_div_metrics.html">get_all_div_metrics()</a></code>

        </dt>
        <dd>Returns all diversity metric options for <code><a href="../reference/calculate_diversity.html">calculate_diversity()</a></code></dd>
      </dl><dl><dt>

          <code><a href="get_all_tcrs.html">get_all_tcrs()</a></code>

        </dt>
        <dd>Returns all of the paired TCRs from all samples in a dataset</dd>
      </dl><dl><dt>

          <code><a href="get_pair_stats.html">get_pair_stats()</a></code>

        </dt>
        <dd>Count the number of single-chains that were paired by each algorithm</dd>
      </dl><dl><dt>

          <code><a href="get_paired_by_read_fraction_range.html">get_paired_by_read_fraction_range()</a></code>

        </dt>
        <dd>Calculate the number and fraction of single chains that were paired within different read fraction ranges</dd>
      </dl><dl><dt>

          <code><a href="summarize_data.html">summarize_data()</a></code>

        </dt>
        <dd>Create a summary table with number of reads and unique alpha/beta chains observed for each sample</dd>
      </dl></div><div class="section level2">

      <h3 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plot_clone_size_across_samples.html">plot_clone_size_across_samples()</a></code>

        </dt>
        <dd>Line plot of clone read fraction across multiple samples</dd>
      </dl><dl><dt>

          <code><a href="plot_clonotype_indices.html">plot_clonotype_indices()</a></code>

        </dt>
        <dd>Stacked bar chart with fractions of reads attributed to the most frequent clonotypes</dd>
      </dl><dl><dt>

          <code><a href="plot_clusters.html">plot_clusters()</a></code>

        </dt>
        <dd>Plot clusters of similar TCRs</dd>
      </dl><dl><dt>

          <code><a href="plot_diversity.html">plot_diversity()</a></code>

        </dt>
        <dd>Bar plot of clonal diversity metrics</dd>
      </dl><dl><dt>

          <code><a href="plot_n_reads.html">plot_n_reads()</a></code>

        </dt>
        <dd>Bar plot of the number of single-chain reads for each sample</dd>
      </dl><dl><dt>

          <code><a href="plot_num_partners.html">plot_num_partners()</a></code>

        </dt>
        <dd>Stacked bar plot of the fraction of single-chains with different numbers of partners for each sample</dd>
      </dl><dl><dt>

          <code><a href="plot_paired.html">plot_paired()</a></code>

        </dt>
        <dd>Stacked bar plot of the number alpha/beta chains paired by each pairing algorithm</dd>
      </dl><dl><dt>

          <code><a href="plot_paired_by_read_fraction_range.html">plot_paired_by_read_fraction_range()</a></code>

        </dt>
        <dd>Bar plot of the fraction of single chains that were paired within different read fraction ranges for each sample.</dd>
      </dl><dl><dt>

          <code><a href="plot_paired_vs_rank.html">plot_paired_vs_rank()</a></code>

        </dt>
        <dd>A stepped plot of the cumulative number of paired (or unpaired) single-chains for the N most frequent single-chains</dd>
      </dl><dl><dt>

          <code><a href="plot_pairs_with_eachother.html">plot_pairs_with_eachother()</a></code>

        </dt>
        <dd>Point plot of read fraction vs. rank for the N most frequent alpha (left, mirrored) and beta (right) chains with lines between alpha and beta chains indicating a pair and a cross indicating an unpaired single-chain</dd>
      </dl><dl><dt>

          <code><a href="plot_ranks.html">plot_ranks()</a></code> <code><a href="plot_ranks.html">rank_plot()</a></code>

        </dt>
        <dd>Line plot of clonotype rank vs. read fraction for each sample</dd>
      </dl><dl><dt>

          <code><a href="plot_read_fraction_vs_pair_status.html">plot_read_fraction_vs_pair_status()</a></code>

        </dt>
        <dd>Alternative point plot of read fraction vs. rank for the N most frequent alpha and beta chains, with a cross indicating an unpaired single-chain</dd>
      </dl><dl><dt>

          <code><a href="plot_sample_overlap.html">plot_sample_overlap()</a></code> <code><a href="plot_sample_overlap.html">sample_overlap()</a></code>

        </dt>
        <dd>Plot the overlap/agreement between samples (in terms of most frequent clones)</dd>
      </dl><dl><dt>

          <code><a href="plot_sample_vs_sample.html">plot_sample_vs_sample()</a></code>

        </dt>
        <dd>Scatterplot of TCR clone read fraction of clones between two samples</dd>
      </dl></div><div class="section level2">

      <h3 id="export-for-external-tools">Export for external tools<a class="anchor" aria-label="anchor" href="#export-for-external-tools"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="create_thimble_df.html">create_thimble_df()</a></code>

        </dt>
        <dd>Convert paired TCRs to a "thimble" data frame for stitching together TCR nucleotide coding sequences with the Stitchr Python package</dd>
      </dl><dl><dt>

          <code><a href="write_stitchr_tsv.html">write_stitchr_tsv()</a></code>

        </dt>
        <dd>Write a tab-separated "thimble" text file for use with Stitchr</dd>
      </dl></div><div class="section level2">

      <h3 id="well-level-data-functions">Well-level data functions<a class="anchor" aria-label="anchor" href="#well-level-data-functions"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="filter_well_data.html">filter_well_data()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Remove rare clones from well-level TIRTLseq data</dd>
      </dl><dl><dt>

          <code><a href="get_well_subset.html">get_well_subset()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Get well names from numerical rows and columns</dd>
      </dl><dl><dt>

          <code><a href="get_wells_from_edges.html">get_wells_from_edges()</a></code>

        </dt>
        <dd>Get well names from top left and bottom right wells</dd>
      </dl><dl><dt>

          <code><a href="load_well_data.html">load_well_data()</a></code>

        </dt>
        <dd>Load well-level TIRTLseq data</dd>
      </dl><dl><dt>

          <code><a href="load_well_data_to_sparse_multi.html">load_well_data_to_sparse_multi()</a></code>

        </dt>
        <dd>Load well-level TIRTLseq data to sparse matrices</dd>
      </dl></div><div class="section level2">

      <h3 id="download-datasets">Download datasets<a class="anchor" aria-label="anchor" href="#download-datasets"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="download_dataset.html">download_dataset()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Download TIRTLseq datasets</dd>
      </dl></div><div class="section level2">

      <h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="params.html">params</a></code>

        </dt>
        <dd>Table of permissible amino acids and V-segments for TCRdist</dd>
      </dl><dl><dt>

          <code><a href="submat.html">submat</a></code>

        </dt>
        <dd>Substitution penalty matrix for TCRdist amino acids and V-segments</dd>
      </dl></div><div class="section level2">
      <h2 id="single-cell-tirtlseq-data">Single-cell TIRTLseq data<a class="anchor" aria-label="anchor" href="#single-cell-tirtlseq-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="clean_scTIRTLseq.html">clean_scTIRTLseq()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Imputes missing alpha and beta chains where possible for single-cell TIRTLseq data</dd>
      </dl><dl><dt>

          <code><a href="summarize_scTIRTLseq.html">summarize_scTIRTLseq()</a></code>
          <span class="badge lifecycle lifecycle-experimental">experimental</span>
        </dt>
        <dd>Summarize single-cell TIRTLseq data</dd>
      </dl></div><div class="section level2">
      <h2 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a></h2>




    </div><div class="section level2">





    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

