<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Find TCRalpha/beta pairs from individual well read counts — run_pairing • TIRTLtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Find TCRalpha/beta pairs from individual well read counts — run_pairing"><meta name="description" content="
This runs the MAD-HYPE and T-SHELL algorithms to find TCRalpha-beta pairs
originating from the same clone."><meta property="og:description" content="
This runs the MAD-HYPE and T-SHELL algorithms to find TCRalpha-beta pairs
originating from the same clone."><meta property="og:image" content="https://nicholasclark.github.io/TIRTLtools/logo.svg"></head><body><p>









  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/TIRTLtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/data_loading.html">Loading TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/plot_examples.html">Plot examples with SJTRC COVID-19 data</a></li>
    <li><a class="dropdown-item" href="../articles/TCRdist.html">Fast TCR similarity calculation with TCRdist GPU</a></li>
    <li><a class="dropdown-item" href="../articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic TIRTLseq data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/summarizing_diversity.html">Estimating TCR Repertoire diversity</a></li>
    <li><a class="dropdown-item" href="../articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Constructing full TCR nucleotide sequences</h6></li>
    <li><a class="dropdown-item" href="../articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Well-level TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Find TCRalpha/beta pairs from individual well read counts</h1>

      <div class="d-none name"><code>run_pairing.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a>
This runs the MAD-HYPE and T-SHELL algorithms to find TCRalpha-beta pairs
originating from the same clone.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_pairing</span><span class="op">(</span></span>
<span>  <span class="va">folder_path</span>,</span>
<span>  <span class="va">folder_out</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"tmp"</span>,</span>
<span>  well_filter_thres <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  min_reads <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_wells <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  well_pos <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  wellset1 <span class="op">=</span> <span class="fu"><a href="get_well_subset.html">get_well_subset</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">16</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">24</span><span class="op">)</span>,</span>
<span>  compute <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  backend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"auto"</span>, <span class="st">"cpu"</span>, <span class="st">"cupy"</span>, <span class="st">"mlx"</span><span class="op">)</span>,</span>
<span>  pval_thres_tshell <span class="op">=</span> <span class="fl">1e-10</span>,</span>
<span>  wij_thres_tshell <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write_extra_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filter_before_top3 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fork <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shared <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-folder-path">folder_path<a class="anchor" aria-label="anchor" href="#arg-folder-path"></a></dt>
<dd><p>the path of the folder with well-level data</p></dd>


<dt id="arg-folder-out">folder_out<a class="anchor" aria-label="anchor" href="#arg-folder-out"></a></dt>
<dd><p>the path of the folder to write results to. The function will create the folder if it does not exist.</p></dd>


<dt id="arg-prefix">prefix<a class="anchor" aria-label="anchor" href="#arg-prefix"></a></dt>
<dd><p>a prefix for the output file names</p></dd>


<dt id="arg-well-filter-thres">well_filter_thres<a class="anchor" aria-label="anchor" href="#arg-well-filter-thres"></a></dt>
<dd><p>wells are removed if they have fewer unique clones than: wellfilter_thres*(Avg. # of unique clones per well).
The default value is 0.75</p></dd>


<dt id="arg-min-reads">min_reads<a class="anchor" aria-label="anchor" href="#arg-min-reads"></a></dt>
<dd><p>minimum number of reads a chain must have in a well to be considered observed
(note: actual minimum is min_reads+1. Default value is 0, i.e. chain must have &gt;= 1 read in a well)</p></dd>


<dt id="arg-min-wells">min_wells<a class="anchor" aria-label="anchor" href="#arg-min-wells"></a></dt>
<dd><p>minimum number of wells a chain must be observed in to be paired.</p></dd>


<dt id="arg-well-pos">well_pos<a class="anchor" aria-label="anchor" href="#arg-well-pos"></a></dt>
<dd><p>the position of the well ID (e.g. "B5") in the file names. For example, files
named "<lab><em><project></project></em>&lt;well_id&gt;_TCRalpha.tsv" would use well_pos=3. (default is 3)</lab></p></dd>


<dt id="arg-wellset-">wellset1<a class="anchor" aria-label="anchor" href="#arg-wellset-"></a></dt>
<dd><p>a vector of wells to use for the pairing</p></dd>


<dt id="arg-compute">compute<a class="anchor" aria-label="anchor" href="#arg-compute"></a></dt>
<dd><p>whether or not to run the pairing algorithms after tabulating and writing pseudobulk data (default TRUE)</p></dd>


<dt id="arg-backend">backend<a class="anchor" aria-label="anchor" href="#arg-backend"></a></dt>
<dd><p>the computing backend to use. The function looks for a GPU and automatically chooses an appropriate backend by default.</p></dd>


<dt id="arg-pval-thres-tshell">pval_thres_tshell<a class="anchor" aria-label="anchor" href="#arg-pval-thres-tshell"></a></dt>
<dd><p>the adjusted p-value threshold for T-SHELL significance (default 1e-10)</p></dd>


<dt id="arg-wij-thres-tshell">wij_thres_tshell<a class="anchor" aria-label="anchor" href="#arg-wij-thres-tshell"></a></dt>
<dd><p>the threshold for the number of wells containing both chains for T-SHELL significance (default &gt;2 wells)</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>whether to print out messages (default TRUE)</p></dd>


<dt id="arg-write-extra-files">write_extra_files<a class="anchor" aria-label="anchor" href="#arg-write-extra-files"></a></dt>
<dd><p>whether to write un-necessary intermediate files (default FALSE)</p></dd>


<dt id="arg-filter-before-top-">filter_before_top3<a class="anchor" aria-label="anchor" href="#arg-filter-before-top-"></a></dt>
<dd><p>whether to filter by loss fraction before extracting top 3 correlation values for T-SHELL (default FALSE)</p></dd>


<dt id="arg-fork">fork<a class="anchor" aria-label="anchor" href="#arg-fork"></a></dt>
<dd><p>whether to "fork" the python process for basilisk (default is NULL, which automatically chooses an appropriate option)</p></dd>


<dt id="arg-shared">shared<a class="anchor" aria-label="anchor" href="#arg-shared"></a></dt>
<dd><p>whether to use a "shared" python process for basilisk (default is NULL, which automatically chooses an appropriate option)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with the TCR-alpha/TCR-beta pairs.</p>
<p>The function also writes three files to "folder_out":</p><ul><li><p>A data frame ("<prefix>_pseudobulk_TRA.tsv") of pseudobulk counts for TCRalpha chains</prefix></p></li>
<li><p>A data frame ("<prefix>_pseudobulk_TRB.tsv") of pseudobulk counts for TCRbeta chains</prefix></p></li>
<li><p>A data frame ("<prefix>_TIRTLoutput.tsv") of TCR-alpha/TCR-beta pairs.</prefix></p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  
</body></html>

