<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>GPU implementation of TCRdist, a distance/similarity metric for pairs of TCRs — TCRdist • TIRTLtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="GPU implementation of TCRdist, a distance/similarity metric for pairs of TCRs — TCRdist"><meta name="description" content="
An efficient, batched version of TCRdist that is compatible with both NVIDIA and Apple Silicon GPUs."><meta property="og:description" content="
An efficient, batched version of TCRdist that is compatible with both NVIDIA and Apple Silicon GPUs."><meta property="og:image" content="https://nicholasclark.github.io/TIRTLtools/logo.svg"></head><body><p>









  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/TIRTLtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/data_loading.html">Loading Paired TCR TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/plot_examples.html">Plot examples with SJTRC COVID-19 data</a></li>
    <li><a class="dropdown-item" href="../articles/pairing_zenodo_data.html">TCR-alpha/beta pairing for COVID-19 related TCRs</a></li>
    <li><a class="dropdown-item" href="../articles/TCRdist.html">Fast TCR similarity calculation with TCRdist GPU</a></li>
    <li><a class="dropdown-item" href="../articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic TIRTLseq data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/summarizing_diversity.html">Estimating TCR Repertoire diversity</a></li>
    <li><a class="dropdown-item" href="../articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Constructing full TCR nucleotide sequences</h6></li>
    <li><a class="dropdown-item" href="../articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Well-level TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>GPU implementation of TCRdist, a distance/similarity metric for pairs of TCRs</h1>

      <div class="d-none name"><code>TCRdist.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a>
An efficient, batched version of TCRdist that is compatible with both NVIDIA and Apple Silicon GPUs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TCRdist</span><span class="op">(</span></span>
<span>  <span class="va">tcr1</span>,</span>
<span>  tcr2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_MAIT <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  params <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  submat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tcrdist_cutoff <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  chunk_size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  print_chunk_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  print_res <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  only_lower_tri <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write_to_tsv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  backend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"auto"</span>, <span class="st">"cpu"</span>, <span class="st">"cupy"</span>, <span class="st">"mlx"</span><span class="op">)</span>,</span>
<span>  fork <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shared <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tcr-">tcr1<a class="anchor" aria-label="anchor" href="#arg-tcr-"></a></dt>
<dd><p>a data frame with one TCR per row. It must have the columns "va", "vb", "cdr3a", and "cdr3b"</p></dd>


<dt id="arg-tcr-">tcr2<a class="anchor" aria-label="anchor" href="#arg-tcr-"></a></dt>
<dd><p>(optional) another data frame of TCRs. If supplied, TCRdist will be calculated
for every combination of one TCR from tcr1 and one TCR from tcr2. Otherwise, it will calculate TCRdist
for every pair of TCRs in tcr1.</p></dd>


<dt id="arg-remove-mait">remove_MAIT<a class="anchor" aria-label="anchor" href="#arg-remove-mait"></a></dt>
<dd><p>whether to remove TCRs from MAIT cells (default is FALSE)</p></dd>


<dt id="arg-params">params<a class="anchor" aria-label="anchor" href="#arg-params"></a></dt>
<dd><p>(optional) a table of valid parameters for amino acids and va/vb segments.
(default is NULL, which uses TIRTLtools::params)</p></dd>


<dt id="arg-submat">submat<a class="anchor" aria-label="anchor" href="#arg-submat"></a></dt>
<dd><p>(optional) a substitution matrix with mismatch penalties for each
combination of amino acids or va/vb segments (default is NULL, which uses TIRTLtools::submat).</p></dd>


<dt id="arg-tcrdist-cutoff">tcrdist_cutoff<a class="anchor" aria-label="anchor" href="#arg-tcrdist-cutoff"></a></dt>
<dd><p>(optional) discard all TCRdist values above this cutoff (default is 90).</p></dd>


<dt id="arg-chunk-size">chunk_size<a class="anchor" aria-label="anchor" href="#arg-chunk-size"></a></dt>
<dd><p>(optional) The chunk size to use in calculation of TCRdist (default 1000). If set at n,
it will calculate pairwise TCRdist for n x n TCRs at once. This may be as high as allowable by GPU memory
(in our testing, a chunk_size of 1000 to 5000 provided the fastest runtime and
chunk_size of over 7500 resulted in memory errors on some GPUs).</p></dd>


<dt id="arg-print-chunk-size">print_chunk_size<a class="anchor" aria-label="anchor" href="#arg-print-chunk-size"></a></dt>
<dd><p>(optional) print a line of output for every n TCRs processed (default 1000)</p></dd>


<dt id="arg-print-res">print_res<a class="anchor" aria-label="anchor" href="#arg-print-res"></a></dt>
<dd><p>(optional) print summary of results (default is TRUE)</p></dd>


<dt id="arg-only-lower-tri">only_lower_tri<a class="anchor" aria-label="anchor" href="#arg-only-lower-tri"></a></dt>
<dd><p>(optional) return one TCRdist value for each pair (like the lower triangle of a symmetric matrix). Default is TRUE.</p></dd>


<dt id="arg-return-data">return_data<a class="anchor" aria-label="anchor" href="#arg-return-data"></a></dt>
<dd><p>(optional) whether to return the output result from the function.
With large data it may be desirable to write the result to disk instead. (default is TRUE, returns output)</p></dd>


<dt id="arg-write-to-tsv">write_to_tsv<a class="anchor" aria-label="anchor" href="#arg-write-to-tsv"></a></dt>
<dd><p>(optional) write the results to a tab-separated file ".tsv" (default is FALSE, does not write .tsv file)</p></dd>


<dt id="arg-backend">backend<a class="anchor" aria-label="anchor" href="#arg-backend"></a></dt>
<dd><p>(optional) the CPU or GPU backend to use (default "auto")</p></dd>


<dt id="arg-fork">fork<a class="anchor" aria-label="anchor" href="#arg-fork"></a></dt>
<dd><p>(optional) a TRUE/FALSE value for whether to "fork" a new Python process for running TCRdist via the "basilisk" package. Default is NULL, which should use choose a safe value based on how the package is loaded.</p></dd>


<dt id="arg-shared">shared<a class="anchor" aria-label="anchor" href="#arg-shared"></a></dt>
<dd><p>(optional) a TRUE/FALSE value for whether to "share" the Python process for running TCRdist via the "basilisk" package. Default is NULL, which should use choose a safe value based on how the package is loaded.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with entries:</p>
<p><code>$TCRdist_df</code> - a data frame with three columns: "node1_0index", "node2_0index", and "TCRdist".
The first two columns contain the indices (0-indexed) of the TCRs for each pair.
The last column contains the TCRdist if it is below the cutoff. The output is sparse in that it only contains
pairs that have TCRdist &lt;= cutoff.</p>
<p><code>$tcr1</code> - a data frame of the TCRs supplied to the function. It contains an additional column
"tcr_index" with the (0-indexed) index of each TCR.</p>
<p><code>$tcr2</code> - a similar data frame for tcr2, if it was supplied.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function calculates pairwise TCRdist (Dash et al., Nature 2017) for a set of TCRs
(or between two sets of TCRs) and returns a sparse output with the TCRdist and indices of all pairs
that have TCRdist less than or equal to a desired cutoff (default cutoff is 90).</p>
<p>The function uses the <code>reticulate</code> package to call a python script that uses <code>cupy</code> (NVIDIA GPUs), <code>mlx</code> (Apple Silicon GPUs),
or <code>numpy</code> (no GPU) to calculate TCRdist efficiently.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="cluster_tcrs.html">cluster_tcrs()</a></code>, <code><a href="plot_clusters.html">plot_clusters()</a></code>, and <code><a href="identify_non_functional_seqs.html">identify_non_functional_seqs()</a></code></p>
<p>Other tcr_similarity:
<code><a href="TCRdist_cpp.html">TCRdist_cpp()</a></code>,
<code><a href="cluster_tcrs.html">cluster_tcrs()</a></code>,
<code><a href="plot_clusters.html">plot_clusters()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/SJTRC_TIRTLseq_minimal"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"TIRTLtools"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sjtrc</span> <span class="op">=</span> <span class="fu"><a href="load_tirtlseq.html">load_tirtlseq</a></span><span class="op">(</span><span class="va">folder</span>,</span></span>
<span class="r-in"><span>  meta_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"marker"</span>, <span class="st">"timepoint"</span>, <span class="st">"version"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span>,</span></span>
<span class="r-in"><span>  chain <span class="op">=</span> <span class="st">"paired"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading files from: /Users/nclark52/git/TIRTLtools/inst/extdata/SJTRC_TIRTLseq_minimal...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2 beta chain pseudo-bulk files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2 paired chain files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loaded 2 files from 2 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.2 seconds</span>
<span class="r-in"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="get_all_tcrs.html">get_all_tcrs</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain<span class="op">=</span><span class="st">"paired"</span>, remove_duplicates <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">=</span> <span class="fu">TCRdist</span><span class="op">(</span><span class="va">df</span>, tcrdist_cutoff <span class="op">=</span> <span class="fl">90</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 384 TCRs with unknown V-segments (1.2%) from a total of 32,164 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 12 TCRs with short CDR3 segments (0.038%) from a total of 31,780 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 13,324 TCRs with non-functional CDR3 amino acid sequences (42%) from a total of 31,768 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtered data frame contains 18,444 TCRs (57%) of original 32,164 TCRs.</span>
<span class="r-in"><span><span class="va">edge_df</span> <span class="op">=</span> <span class="va">result</span><span class="op">[[</span><span class="st">'TCRdist_df'</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="op">)</span> <span class="co">### table of TCRdist values &lt;= cutoff</span></span></span>
<span class="r-in"><span><span class="va">node_df</span> <span class="op">=</span> <span class="va">result</span><span class="op">[[</span><span class="st">'tcr1'</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="op">)</span> <span class="co">### table of input data with indices</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">edge_df</span> <span class="co">## sparse 3-column output: node1, node2, TCRdist</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        node1_0index node2_0index TCRdist</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               &lt;int&gt;        &lt;int&gt;   &lt;int&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1:          158          157       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2:          286          284      84</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3:          287          283      89</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4:          288          284      89</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     5:          289          285      89</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ---                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18422:        18434        18073       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18423:        18437        18406      72</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18424:        18440        18241      90</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18425:        18440        18324      66</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18426:        18440        18422      84</span>
<span class="r-in"><span><span class="co">## note that indices start at 0 and are found in node_df$tcr_index</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">node_df</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">select</span><span class="op">(</span><span class="va">tcr_index</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">mutate</span><span class="op">(</span>alpha_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">alpha_nuc</span>, <span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span>, <span class="st">"..."</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>         beta_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">beta_nuc</span>, <span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span>, <span class="st">"..."</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in mutate(., alpha_nuc = paste(substr(alpha_nuc, 0, 20), "...",     sep = ""), beta_nuc = paste(substr(beta_nuc, 0, 20), "...",     sep = "")):</span> could not find function "mutate"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  
</body></html>

