<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Parallelized C++ implementation of TCRdist (no GPU required) — TCRdist_cpp • TIRTLtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Parallelized C++ implementation of TCRdist (no GPU required) — TCRdist_cpp"><meta name="description" content="
This is an alternative to the GPU version of TCRdist that is still very fast for
large datasets (tens of thousands of TCRs). It is written in C++ and will
run in parallel across available CPU cores."><meta property="og:description" content="
This is an alternative to the GPU version of TCRdist that is still very fast for
large datasets (tens of thousands of TCRs). It is written in C++ and will
run in parallel across available CPU cores."><meta property="og:image" content="https://nicholasclark.github.io/TIRTLtools/logo.svg"></head><body><p>









  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/TIRTLtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/data_loading.html">Loading Paired TCR TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/plot_examples.html">Plot examples with SJTRC COVID-19 data</a></li>
    <li><a class="dropdown-item" href="../articles/pairing_zenodo_data.html">TCR-alpha/beta pairing for COVID-19 related TCRs</a></li>
    <li><a class="dropdown-item" href="../articles/TCRdist.html">Fast TCR similarity calculation with TCRdist GPU</a></li>
    <li><a class="dropdown-item" href="../articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic TIRTLseq data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/summarizing_diversity.html">Estimating TCR Repertoire diversity</a></li>
    <li><a class="dropdown-item" href="../articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Constructing full TCR nucleotide sequences</h6></li>
    <li><a class="dropdown-item" href="../articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Well-level TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Parallelized C++ implementation of TCRdist (no GPU required)</h1>

      <div class="d-none name"><code>TCRdist_cpp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a>
This is an alternative to the GPU version of TCRdist that is still very fast for
large datasets (tens of thousands of TCRs). It is written in C++ and will
run in parallel across available CPU cores.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TCRdist_cpp</span><span class="op">(</span><span class="va">tcr1</span>, tcr2 <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tcr-">tcr1<a class="anchor" aria-label="anchor" href="#arg-tcr-"></a></dt>
<dd><p>a data frame with one TCR per row. It must have the columns "va", "vb", "cdr3a", and "cdr3b"</p></dd>


<dt id="arg-tcr-">tcr2<a class="anchor" aria-label="anchor" href="#arg-tcr-"></a></dt>
<dd><p>(optional) another data frame of TCRs. If supplied, TCRdist will be calculated
for every combination of one TCR from tcr1 and one TCR from tcr2. Otherwise, it will calculate TCRdist
for every pair of TCRs in tcr1.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list with two objects (or three if tcr2 is not null):</p><ul><li><p>matrix - a matrix of TCRdist values</p></li>
<li><p>tcr1 - the input matrix tcr1, after pre-processing and removing unacceptable TCRs</p></li>
<li><p>tcr2 (if supplied) - the input matrix tcr2, after pre-processing and removing unacceptable TCRs</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This version of TCRdist is currently less feature-rich than <code><a href="TCRdist.html">TCRdist()</a></code> and returns
a dense matrix as output. It does not yet allow for sparse output or writing output directly to a file.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other tcr_similarity:
<code><a href="TCRdist.html">TCRdist()</a></code>,
<code><a href="cluster_tcrs.html">cluster_tcrs()</a></code>,
<code><a href="plot_clusters.html">plot_clusters()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/SJTRC_TIRTLseq_minimal"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"TIRTLtools"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sjtrc</span> <span class="op">=</span> <span class="fu"><a href="load_tirtlseq.html">load_tirtlseq</a></span><span class="op">(</span><span class="va">folder</span>,</span></span>
<span class="r-in"><span>  meta_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"marker"</span>, <span class="st">"timepoint"</span>, <span class="st">"version"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span>,</span></span>
<span class="r-in"><span>  chain <span class="op">=</span> <span class="st">"paired"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading files from: /Users/nclark52/git/TIRTLtools/inst/extdata/SJTRC_TIRTLseq_minimal...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2 beta chain pseudo-bulk files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2 paired chain files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loaded 2 files from 2 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.2 seconds</span>
<span class="r-in"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="get_all_tcrs.html">get_all_tcrs</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain<span class="op">=</span><span class="st">"paired"</span>, remove_duplicates <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">=</span> <span class="fu">TCRdist_cpp</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 384 TCRs with unknown V-segments (1.2%) from a total of 32,164 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 12 TCRs with short CDR3 segments (0.038%) from a total of 31,780 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 13,324 TCRs with non-functional CDR3 amino acid sequences (42%) from a total of 31,768 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 812 MAIT TCRs (4.4%) from a total of 18,444 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtered data frame contains 17,632 TCRs (55%) of original 32,164 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 384 TCRs with unknown V-segments (1.2%) from a total of 32,164 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 12 TCRs with short CDR3 segments (0.038%) from a total of 31,780 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 13,324 TCRs with non-functional CDR3 amino acid sequences (42%) from a total of 31,768 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 812 MAIT TCRs (4.4%) from a total of 18,444 TCRs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtered data frame contains 17,632 TCRs (55%) of original 32,164 TCRs.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">matrix</span></span></span>
<span class="r-in"><span><span class="va">node_df</span> <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">tcr1</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1] [,2] [,3] [,4] [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]   NA   NA   NA   NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  338   NA   NA   NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  287  270   NA   NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,]  374  320  293   NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]  382  372  386  387   NA</span>
<span class="r-in"><span><span class="co">## note: TCRdist is symmetric. Returned matrix contains only lower-triangular values</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">node_df</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>alpha_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">alpha_nuc</span>, <span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span>, <span class="st">"..."</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>         beta_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">beta_nuc</span>, <span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span>, <span class="st">"..."</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in mutate(., alpha_nuc = paste(substr(alpha_nuc, 0, 20), "...",     sep = ""), beta_nuc = paste(substr(beta_nuc, 0, 20), "...",     sep = "")):</span> could not find function "mutate"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  
</body></html>

