<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Plot examples with SJTRC COVID-19 data • TIRTLtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Plot examples with SJTRC COVID-19 data">
</head>
<body>
<p>








  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TIRTLtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/TIRTLtools.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/data_loading.html">Loading TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/plot_examples.html">Plot examples with SJTRC COVID-19 data</a></li>
    <li><a class="dropdown-item" href="../articles/TCRdist.html">Fast TCR similarity calculation with TCRdist GPU</a></li>
    <li><a class="dropdown-item" href="../articles/structure_in_R.html">(coming soon) The structure of TIRTLseq data in R</a></li>
    <li><a class="dropdown-item" href="../articles/working_with_tirtlseq_data_in_R.html">(coming soon) Working with TIRTLseq data in R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic TIRTLseq data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/summarizing_diversity.html">(coming soon) Summarizing TCR diversity</a></li>
    <li><a class="dropdown-item" href="../articles/SJTRC_identifying_covid_epitopes.html">(coming soon) Identifying SARS-CoV-2 epitopes from TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Constructing full TCR nucleotide sequences</h6></li>
    <li><a class="dropdown-item" href="../articles/stitchr.html">(coming soon) Create output for Stitchr</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/scTIRTLseq_data.html">(coming soon) Working with single-cell TIRTLseq data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Well-level TIRTLseq analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/well_level_data.html">(coming soon) Working with well-level TIRTLseq data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicholasClark/TIRTLtools"><span class="fa fa-github fa-lg"></span> Github</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Plot examples with SJTRC COVID-19 data</h1>
            
      

      <div class="d-none name"><code>plot_examples.Rmd</code></div>
    </div>

    
    
<p>This vignette loads paired TCR TIRTLseq data from the St. Jude
Tracking Study of Immune Responses Associated with COVID-19 (SJTRC) and
shows examples of the package’s plotting functions.</p>
<div class="section level2">
<h2 id="load-the-package">Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nicholasclark.github.io/TIRTLtools/">TIRTLtools</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in system("nvidia-smi", intern = TRUE, ignore.stderr = TRUE) : </span></span>
<span><span class="co">##   error in running command</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-sjtrc-covid-19-data">Load the SJTRC COVID-19 data<a class="anchor" aria-label="anchor" href="#load-the-sjtrc-covid-19-data"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/SJTRC_TIRTL_seq_longitudinal"</span>, package <span class="op">=</span> <span class="st">"TIRTLtools"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">folder</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "cd4_tp1_v2_pseudobulk_TRA.tsv.gz" "cd4_tp1_v2_pseudobulk_TRB.tsv.gz"</span></span>
<span><span class="co">##  [3] "cd4_tp1_v2_TIRTLoutput.tsv.gz"    "cd4_tp2_v2_pseudobulk_TRA.tsv.gz"</span></span>
<span><span class="co">##  [5] "cd4_tp2_v2_pseudobulk_TRB.tsv.gz" "cd4_tp2_v2_TIRTLoutput.tsv.gz"   </span></span>
<span><span class="co">##  [7] "cd4_tp3_v2_pseudobulk_TRA.tsv.gz" "cd4_tp3_v2_pseudobulk_TRB.tsv.gz"</span></span>
<span><span class="co">##  [9] "cd4_tp3_v2_TIRTLoutput.tsv.gz"    "cd8_tp1_v2_pseudobulk_TRA.tsv.gz"</span></span>
<span><span class="co">## [11] "cd8_tp1_v2_pseudobulk_TRB.tsv.gz" "cd8_tp1_v2_TIRTLoutput.tsv.gz"   </span></span>
<span><span class="co">## [13] "cd8_tp2_v2_pseudobulk_TRA.tsv.gz" "cd8_tp2_v2_pseudobulk_TRB.tsv.gz"</span></span>
<span><span class="co">## [15] "cd8_tp2_v2_TIRTLoutput.tsv.gz"    "cd8_tp3_v2_pseudobulk_TRA.tsv.gz"</span></span>
<span><span class="co">## [17] "cd8_tp3_v2_pseudobulk_TRB.tsv.gz" "cd8_tp3_v2_TIRTLoutput.tsv.gz"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sjtrc</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_tirtlseq.html">load_tirtlseq</a></span><span class="op">(</span><span class="va">folder</span>, meta_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"marker"</span>, <span class="st">"timepoint"</span>, <span class="st">"version"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 65.148 sec elapsed</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="process-the-data">Process the data<a class="anchor" aria-label="anchor" href="#process-the-data"></a>
</h2>
<p>When we initially load the data, the data frames lack some
information that is needed for plotting functions.</p>
<p>The <code><a href="../reference/TIRTL_process.html">TIRTL_process()</a></code> function runs 3 other package
functions to add this information to the data frames:</p>
<ul>
<li>
<code><a href="../reference/add_single_chain_data.html">add_single_chain_data()</a></code> - adds single-chain read
counts/fractions to the paired TCR data frames</li>
<li>
<code><a href="../reference/identify_paired.html">identify_paired()</a></code> - adds a column to pseudobulk data
indicating which single-chains were paired</li>
<li>
<code><a href="../reference/identify_non_functional_seqs.html">identify_non_functional_seqs()</a></code> - adds columns to paired
data frames indicating whether CDR3A/B amino acid sequences are
functional (not containing stop codons or frameshifts)</li>
</ul>
<p>If <code>clean = TRUE</code> is specified, the
<code><a href="../reference/clean_pairs.html">clean_pairs()</a></code> function is also run before the preceding
functions. This function removes excess pairs for alpha and beta chains.
Biologically, we expect that each alpha is paired with at most one beta
and that each beta is paired with at most two alphas in a clone. Here we
use <code>clean = FALSE</code> to show how the data looks as-is.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sjtrc</span> <span class="op">=</span> <span class="fu"><a href="../reference/TIRTL_process.html">TIRTL_process</a></span><span class="op">(</span><span class="va">sjtrc</span>, clean <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Adding single-chain data to paired dataframe for sample 1</span></span>
<span><span class="co">## Adding single-chain data to paired dataframe for sample 2</span></span>
<span><span class="co">## Adding single-chain data to paired dataframe for sample 3</span></span>
<span><span class="co">## Adding single-chain data to paired dataframe for sample 4</span></span>
<span><span class="co">## Adding single-chain data to paired dataframe for sample 5</span></span>
<span><span class="co">## Adding single-chain data to paired dataframe for sample 6</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 1</span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 2</span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 3</span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 4</span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 5</span></span>
<span><span class="co">## Annotating data with pairing status by MAD-HYPE and T-SHELL algorithms for sample 6</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="longitudinal-plots-for-individual-tcrs-or-groups-of-tcrs">Longitudinal plots for individual TCRs or groups of TCRs<a class="anchor" aria-label="anchor" href="#longitudinal-plots-for-individual-tcrs-or-groups-of-tcrs"></a>
</h2>
<p>For CD8-selected T-cells, we get the nucleotide sequences for the 5
most frequent beta chains for timepoints 1 and 2 and show their
frequencies across all samples and timepoints.</p>
<p>We can see they are highly frequent at CD8 timepoint 3, but not in
any of the CD4 samples.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_clones1</span> <span class="op">=</span> <span class="va">sjtrc</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cd8_tp1_v2</span><span class="op">$</span><span class="va">beta</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">readFraction</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract2</a></span><span class="op">(</span><span class="st">"targetSequences"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">top_clones2</span> <span class="op">=</span> <span class="va">sjtrc</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cd8_tp2_v2</span><span class="op">$</span><span class="va">beta</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">readFraction</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract2</a></span><span class="op">(</span><span class="st">"targetSequences"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_clone_size_across_samples.html">plot_clone_size_across_samples</a></span><span class="op">(</span><span class="va">sjtrc</span>, clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">top_clones1</span>, <span class="va">top_clones2</span><span class="op">)</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" width="700"></p>
<p>We can also plot the proportion of reads occupied by the 10 most
frequent clones, or the top 100, or 1000, etc for each sample.</p>
<p>Here we can see that the top CD8 clones occupy more of the repertoire
than their CD4 counterparts.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_clonotype_indices.html">plot_clonotype_indices</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" width="700"></p>
<p>A good robust measure of repertoire diversity is the number of clones
needed to make up half of the total reads (d50).</p>
<p>A higher d50 means a less clonal repertoire and a lower d50 means a
repertoire dominated by a small number of clones.</p>
<p>We see that the CD8 repertoire is more clonal than the CD4 and at
timepoint 2 it is the most clonal.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">div</span> <span class="op">=</span> <span class="fu"><a href="../reference/calculate_diversity.html">calculate_diversity</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain <span class="op">=</span> <span class="st">"beta"</span>, metrics <span class="op">=</span> <span class="st">"d50"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## -- Calculating diversity indices for sample 1 of 6.</span></span>
<span><span class="co">## -- Calculating diversity indices for sample 2 of 6.</span></span>
<span><span class="co">## -- Calculating diversity indices for sample 3 of 6.</span></span>
<span><span class="co">## -- Calculating diversity indices for sample 4 of 6.</span></span>
<span><span class="co">## -- Calculating diversity indices for sample 5 of 6.</span></span>
<span><span class="co">## -- Calculating diversity indices for sample 6 of 6.</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span><span class="va">div</span>, metric <span class="op">=</span> <span class="st">"d50"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" width="700">
The package offers a large number of other diversity metrics, including
Simpson, inverse Simpson, Shannon, richness, etc.</p>
<p>For a full list of the available metrics, run
<code><a href="../reference/get_all_div_metrics.html">get_all_div_metrics()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_all_div_metrics.html">get_all_div_metrics</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "simpson"        "gini"           "gini.simpson"   "inv.simpson"   </span></span>
<span><span class="co">##  [5] "shannon"        "berger.parker"  "richness"       "d50"           </span></span>
<span><span class="co">##  [9] "dXX"            "renyi"          "hill"           "top10fraction" </span></span>
<span><span class="co">## [13] "top100fraction" "topNfraction"</span></span></code></pre>
<p>We can also plot the total number of alpha and beta chain reads for
each sample.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_n_reads.html">plot_n_reads</a></span><span class="op">(</span><span class="va">sjtrc</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-8-1.png" class="r-plt" width="700"></p>
<p>We can also check how many alpha and beta chains were paired for each
sample and by which algorithm.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_paired.html">plot_paired</a></span><span class="op">(</span><span class="va">sjtrc</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-9-1.png" class="r-plt" width="700"></p>
<p>For quality control purposes, for the chains that are paired, we can
plot how many partners each was assigned.</p>
<p>We expect alpha chains to have at most one beta chain partner and
beta chains to have at most two alpha chain partners. Higher numbers of
partners could be indicative of problems with the pairing. Sometimes
these extra partners indicate mispairings and sometimes they are
sequencing/PCR errors of the true partner.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_num_partners.html">plot_num_partners</a></span><span class="op">(</span><span class="va">sjtrc</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" width="700"></p>
<p>We can also check the percentage of chains that are paired within
each read fraction range. We would expect good pairing for highly
frequent chains and worse pairing for rare chains.</p>
<p>Here we see that beta chains that are more frequent than one in
ten-thousand (0.0001, darkest three bars) are paired very well in all
samples. For rarer chains, the fraction paired drops off and almost none
of the chains in the one-in-one-hundred-thousand to one-in-one-million
range (light green, [1e-06,1e-05]) are paired.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_paired_by_read_fraction_range.html">plot_paired_by_read_fraction_range</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" width="700"></p>
<p>We might expect that our algorithms (at least T-SHELL) are able to
pair the majority of the most frequent clones. We can check how often
our pairing algorithms fail for the top clones with a stepped line-chart
that tracks how many of the top clones we fail to pair for a given
sample.</p>
<p>For this CD8 sample at timepoint 1, we see that the T-SHELL algorithm
is able to pair almost all of the top 100 clones, but the MAD-HYPE
algorithm is only able to pair about half. This is expected since
MAD-HYPE calls pairs based on occurrence patterns and therefore has a
very difficult time pairing clones that are found in all or almost all
wells.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_paired_vs_rank.html">plot_paired_vs_rank</a></span><span class="op">(</span><span class="va">sjtrc</span>, sample <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-12-1.png" class="r-plt" width="700">
Another way to visualize the pairing of the top clones is to plot their
rank from left to right with their read fraction on the y-axis, with
points colored by the pairing status. Here, the unpaired clones are
shown by green dots with a cross. Clones paired by only the T-SHELL
algorithm are shown in yellow. We can see again that T-SHELL works well
for the most frequent clones while the MAD-HYPE algorithm has trouble
with clones that are more frequent than one in a thousand (10^-3).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_read_fraction_vs_pair_status.html">plot_read_fraction_vs_pair_status</a></span><span class="op">(</span><span class="va">sjtrc</span>, sample <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-13-1.png" class="r-plt" width="700">
We might also expect that the most frequent alpha clones are paired to
the most frequent beta clones and vice versa.</p>
<p>This next plot shows the same information as the last one, but with
both alpha and beta clones on the same graph (the top alpha clones are
flipped to the left side). Lines between alpha clones (left) and beta
clones (right) indicate pairs.</p>
<p>We can see that most of the top 100 alpha clones have a partner in
the top 100 beta clones, which makes sense.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pairs_with_eachother.html">plot_pairs_with_eachother</a></span><span class="op">(</span><span class="va">sjtrc</span>, sample <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-14-1.png" class="r-plt" width="700"></p>
<p>We can also plot the rank versus read fraction for all clones and
compare the shapes of these curves among samples.</p>
<p>We can see that the most frequent beta chains from the CD4 samples
have a frequency of around 1 in 1000 (10^-3) while the most frequent
clones from CD8 samples have a frequency of over 1%.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ranks.html">plot_ranks</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-15-1.png" class="r-plt" width="700"></p>
<p>As another quality control check, we can plot how many of the top
clones from each sample are shared in other samples.</p>
<p>Here we plot the overlap of the top 200 clones from each sample and
see that there is significant sharing within CD4 and CD8 samples, but
not between them, which is expected.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_sample_overlap.html">plot_sample_overlap</a></span><span class="op">(</span><span class="va">sjtrc</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-16-1.png" class="r-plt" width="700"></p>
<p>We may also want to find whether clones are expanding or contracting
in-between timepoints. Here, the CD8 clones that are expanding from
timepoint 1 to 2 are shown in orange and those that are contracting are
shown in green.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_sample_vs_sample.html">plot_sample_vs_sample</a></span><span class="op">(</span><span class="va">sjtrc</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cd8_tp1_v2</span>, <span class="va">sjtrc</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cd8_tp2_v2</span>, chain <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_examples_files/figure-html/unnamed-chunk-17-1.png" class="r-plt" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/NicholasClark/" class="external-link">Nicholas Clark</a>, <a href="https://scholar.google.com/citations?user=oiypK_UAAAAJ" class="external-link">Mikhail Pogorelyy</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  
</body>
</html>
